### PR TITLE
Add [JsonConstructor] and support for deserializing with parameterized ctors

### LINKED ISSUE #29895: JsonSerializer support for immutable classes and structs.
A common pattern is to make data objects immutable for many different reasons.
For example `Point`:

```csharp
public class Point
{
    int X { get; }

    int Y { get; }

    public Point(int x, int y) => (X, Y) = (x, y);

    public void Deconstruct(out int x, out int y) => (x, y) = (X, Y);
}
```

It would be very helpful if `JsonSerializer` supported immutable `class`es/`struct`s like that, especially since Newtonsoft Json.NET supports deserialization through the constructor.

However there are a few issues surrounding immutable types:

 - Constructor parameter names are usually `camelCase` while Properties are `PascalCase`. One way to solve this is by setting `JsonSerializerOptions.PropertyNameCaseInsensitive` in to `true`. Another way is to use the [`Deconstruct`-pattern](https://docs.microsoft.com/en-us/dotnet/csharp/deconstruct#deconstructing-user-defined-types) introduced in C# 7.0, instead of Properties for serialization.

 - There may be several constructors (and `Deconstruct` methods). So it's not always clear which to use. Newtonsoft Json.NET uses a [JsonConstructorAttribute.htm](https://www.newtonsoft.com/json/help/html/JsonConstructorAttribute.htm), which has the disadvantage that the data objects need to know about serialization.

- Constructor parameters, properties and `Deconstruct` method parameters might not match up, possibly leading to confusing situations where deserializing a previously serialized object does not work.

- Immutable structs always still have a default parameterless constructor.

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/runtime/issues/29895.

Implements specification in https://github.com/dotnet/runtime/pull/33095.

#### Performance notes

Benchmarks: https://github.com/dotnet/performance/pull/1233.

TLDR

- There's a ~4% regression for existing benchmarks when deserializing type graphs with numerous parameterless ctors (specifically `IndexViewModel` and `MyEventListerViewModel` benchmarks). There's also a ~300B increase in allocs for these cases. This is due to new state added to the `ReadStack` to support this feature, and a minimal check when pushing/popping stack frames to support fetching ctor arg state on continuation and/or scenarios involving nested objects with parameterized ctors.
- There's a minimum alloc cost of 200B when deserializing objects with parameterized ctors, compared to deserializing the same object using parameteless ctors (baseline), assuming all members are deserialized in both cases. Total alloc increase depends on the number of ctor args, and how many of them are value types.
- Deserialization of "immutable" objects (where every JSON property maps with a ctor arg) with parameterized ctors is 0-15% slower than baseline. Smaller object graphs are slower due to the overhead of setting up a cache for the arguments.
- Deserialization of mutable objects (where some JSON properties map to object members) with is 0-67% slower than baseline. The slowest cases are when there are numerous object members to deserialize. This is due to re-reading JSON properties after ctor args are parsed and the object is constructed.

### CODE REVIEW

**Comment by steveharter** [comment]
> Benchmarks: dotnet/performance#1233.

Does this degrade perf on the existing benchmarks?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 1]
```diff
@@ -15,26 +15,13 @@ namespace System.Text.Json
...
-        private const int PropertyNameCountCacheThreshold = 64;
-
-        // All of the serializable properties on a POCO (except the optional extension property) keyed on property name.
-        public volatile Dictionary<string, JsonPropertyInfo>? PropertyCache;
+        public delegate object? ConstructorDelegate();
 
-        // All of the serializable properties on a POCO including the optional extension property.
-        // Used for performance during serialization instead of 'PropertyCache' above.
-        public volatile JsonPropertyInfo[]? PropertyCacheArray;
+        public delegate TTypeToConvert ParameterizedConstructorDelegate<TTypeToConvert>(object[] arguments);
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: Not a fan of `TTypeToConverter` since `T` means "Type" so there's a double "type". I'd be fine with just `T` or perhaps `TValue`.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonParameterInfo.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Reflection;
+using System.Text.Json.Serialization;
+
+namespace System.Text.Json
+{
+    [DebuggerDisplay("ParameterInfo={ParameterInfo}")]
+    internal abstract class JsonParameterInfo
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonParameterInfo.cs]: Please add a comment about what this class does. Is "JsonConstructorParameterInfo" better? Or could this be used in non-ctor cases in the future?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 3]
```diff
@@ -214,9 +179,67 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
...
+
+            // It is invalid for the extension data property to bind with a constructor argument.
+            if (DataExtensionProperty != null &&
+                parameterCache.ContainsKey(DataExtensionProperty.NameAsString!))
+            {
+                ThrowHelper.ThrowInvalidOperationException_ExtensionDataCannotBindToCtorParam(DataExtensionProperty.PropertyInfo!, Type, constructorInfo);
+            }
+
+            ParameterCache = parameterCache;
+            ParameterCount = parameters.Length;
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: I don't think this is thread-safe. `ParameterCache` may be used by another thread before `ParameterCount` is set. Do we actually need `ParameterCount` (can we just use `ParameterCache.Count`)?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs [hunk 4]
```diff
@@ -86,14 +117,22 @@ public bool IsProcessingEnumerable()
         public void Reset()
         {
             AddMethodDelegate = null;
+            ConstructorParameterIndex = 0;
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs]: This is a lot of extra state that is not obvious it is tied to ctor parameters. Can we either prefix these with "Ctor_" or something similar, or perhaps move to a new "ConstructorParameters" struct and reference that struct from here?

**Comment by layomia** [comment]
Barring benchmark randomness/error, the only regression is in cases where we have numerous objects (with parameterless ctors) being deserialized, as in the `MyEventsListerViewModel` benchmark which has ~360 objects to be deserialized.

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| ReadJson_MyEventsListerViewModel>.Deseriali |      1.18 |        308692.00 |        363507.78 |         |
| ReadJson_MyEventsListerViewModel>.Deseriali |      1.12 |        375623.49 |        419363.35 |         |
| ReadJson_MyEventsListerViewModel>.Deseriali |      1.12 |        318397.17 |        355241.78 |         |
| ReadJson_Int32>.DeserializeFromString       |      1.07 |           163.71 |           175.62 |         |
| ReadJson_IndexViewModel>.DeserializeFromStr |      1.07 |         29787.77 |         31724.99 |         |
| ReadJson_LargeStructWithProperties>.Deseria |      1.06 |          1146.51 |          1214.22 |         |
| WriteJso_<LoginViewModel>.SerializeToString |      1.06 |           383.82 |           406.37 |         |
| ReadJso_<Int32>.DeserializeFromStream       |      1.06 |           309.70 |           327.77 |         |
| ReadJson_SimpleStructWithProperties>.Deseri |      1.06 |           550.04 |           581.45 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| WriteJson_Location>.SerializeObjectProperty |      1.06 |          1121.74 |          1057.36 |         |

As no changes were made to the post warm-up code paths which handle the deserialization of objects parameterless ctors, I believe the regression is due to the extra state on `ReadStackFrame`, and having additional state to clear after every object is deserialized.

I'll experiment with clearing the extra state only at the start of the deserialization of objects with parameterized ctors and see if this reverts the regression.

[after.zip](https://github.com/dotnet/runtime/files/4325995/after.zip)
[before.zip](https://github.com/dotnet/runtime/files/4325996/before.zip)

File: src/libraries/System.Text.Json/src/Resources/Strings.resx [hunk 5]
```diff
@@ -482,10 +482,25 @@
   <data name="MetadataInvalidPropertyWithLeadingDollarSign" xml:space="preserve">
     <value>Properties that start with '$' are not allowed on preserve mode, either escape the character or turn off preserve references by setting ReferenceHandling to ReferenceHandling.Default.</value>
   </data>
+  <data name="MultipleMembersBindWithConstructorParameter" xml:space="preserve">
+    <value>Members '{0}' and '{1}' on type '{2}' cannot both bind with parameter '{3}' in constructor '{4}' on deserialization.</value>
+  </data>
+  <data name="ConstructorParamIncompleteBinding" xml:space="preserve">
+    <value>Each parameter in constructor '{0}' on type '{1}' must bind to an object member on deserialization. </value>
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/src/Resources/Strings.resx]: The original design was that the property naming policy would be used to determine a JSON name for the arg. The resolution at the spec review was that we should throw if there's an incomplete binding. The reason is so we have clearly defined semantics of how properties should bind with constructor parameters. See:

>If a constructor parameter does not match with a property, InvalidOperationException will be thrown if deserialization is attempted.

in https://github.com/dotnet/runtime/blob/a246af209edb24fa85fb9fe2eed7b3f0b9e8947c/src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md#constructor-parameters-that-bind-to-object-properties-will-use-their-json-property-names-for-deserialization.

A default value will be used if there's no JSON that maps to a constructor parameter.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectConverterFactory.cs [hunk 6]
```diff
@@ -21,16 +22,119 @@ public override bool CanConvert(Type typeToConvert)
...
+                    ctorWithAttribute = constructor;
+                }
+                else if (publicParameterlessCtor == null && constructor.GetParameters().Length == 0)
+                {
+
+                    publicParameterlessCtor = constructor;
+                }
+            }
+
+            // Now, check non-public ctors.
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectConverterFactory.cs]: Yes, I left it in for some time in case there was any pushback. I've removed the support now.

I left  in logic to throw if there are multiple ctors with the `JsonConstructor` attribute (even if one or more are non-public), to avoid first-ctor-with-attribute-wins semantics if we add non-public support in the future (that would be to avoid breaking behavior where we go from not throwing to throwing.)

File: src/libraries/System.Text.Json/tests/Serialization/Object.ReadTests.cs [hunk 7]
```diff
@@ -6,6 +6,7 @@
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.IO;
+using Newtonsoft.Json;
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/tests/Serialization/Object.ReadTests.cs]: Ah, I think I added it when doing some comparison. Thanks

**Comment by layomia** [comment]
> About how much faster is it to remember the reader state instead of just doing a full re-read on the second pass?

- for "immutable" objects where all the JSON properties map to ctor args, it's about the same
- for small objects (2-4 properties) where about half of the JSON properties map to ctor args, ~25% faster
- for objects with ~8 properties where about half of the JSON properties map to ctor args, ~11% faster

[current.zip](https://github.com/dotnet/runtime/files/4327916/current.zip)
[two_pass.zip](https://github.com/dotnet/runtime/files/4327917/two_pass.zip)

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs [hunk 8]
```diff
@@ -0,0 +1,694 @@
...
+                if (jsonPropertyInfo.JsonPropertyName == null)
+                {
+                    propertyNameArray = unescapedPropertyName.ToArray();
+                }
+                else
+                {
+                    propertyNameArray = null;
+                }
+
+                state.Current.FoundProperties![state.Current.FoundPropertyCount++] = (
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs]: Replied in https://github.com/dotnet/runtime/pull/33444#issuecomment-598546371 so I could upload raw benchmarks.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs [hunk 9]
```diff
@@ -0,0 +1,694 @@
...
+                    state.Current.FoundProperties =
+                        ArrayPool<ValueTuple<
+                            JsonPropertyInfo,
+                            JsonReaderState,
+                            long,
+                            byte[]?>>.Shared.Rent(PropertyNameCountCacheThreshold);
+                }
+                else  if (state.Current.FoundPropertyCount == state.Current.FoundProperties!.Length)
+                {
+                    // Case where we can't fit all the JSON properties in the rented pool, we have to grow.
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs]: I'll use that to get the initial/target size, but still need the grow logic for the rare cases where there are duplicate properties in the JSON:

```C#
[Fact]
private static void NumerousPropertiesWork()
{
    StringBuilder sb = new StringBuilder();
    sb.Append("{");
    sb.Append(@"""X"":1,");
    sb.Append(@"""Y"":2,");

    for (int i = 0; i < 65; i++)
    {
        sb.Append($@"""Z"":{i},");
    }

    sb.Append(@"""Z"":66");
    sb.Append("}");

    string json = sb.ToString();

    var point = JsonSerializer.Deserialize<Point_With_Property>(json);
    Assert.Equal(1, point.X);
    Assert.Equal(2, point.Y);
    Assert.Equal(66, point.Z);
}
```

The alternatives are to check for duplicates before caching (slow down perf), or use another data structure e.g. `Dictionary<,>` (we'll allocate).

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: Yup I agree. I had changed it elsewhere and forgot it here. Thanks.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonParameterInfo.cs (hunk 2 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonParameterInfo.cs]: Added.

> Or could this be used in non-ctor cases in the future?

Yes. For this reason I prefer to leave it as "JsonParameterInfo".

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs (hunk 3 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: I believe it is.

`parameters` here is the parameter list for the constructor, which is constant. `ParameterCache` only contains parameters which map to properties, so the count values might be different.

When these count values are different, this means that not all parameters mapped to properties, so we throw an exception if deserialization is attempted: 

https://github.com/dotnet/runtime/blob/9202f23223310cf11a9f39b177e885cda699b72c/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs#L79-L82

I've added some comments.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs (hunk 4 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs]: I've moved them to a new struct.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs [hunk 10]
```diff
@@ -0,0 +1,694 @@
...
+                    reader.BytesConsumed,
+                    propertyNameArray);
+
+                reader.Skip();
+            }
+            else
+            {
+                reader.Read();
+
+                if (state.Current.JsonClassInfo.DataExtensionIsObject)
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs]: With this implementation, reading the extension data alongside the properties would mean caching the reader state for each extra property seen which increases the likelyhood of having to grow the `FoundProperties(Async)` cache (related to the conversation in https://github.com/dotnet/runtime/pull/33444#discussion_r391712806).

Also, it would mean more `if` checks when reading properties after the object is constructed (on the second pass), which would slow down perf when there is no extension data property.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs [hunk 11]
```diff
@@ -0,0 +1,694 @@
...
+            state.Current.UseExtensionProperty = useExtensionProperty;
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void HandleProperty(ref ReadStack state, ref Utf8JsonReader reader, ReadOnlySpan<byte> unescapedPropertyName, JsonPropertyInfo jsonPropertyInfo, JsonSerializerOptions options)
+        {
+            bool useExtensionProperty;
+            // Determine if we should use the extension property.
+            if (jsonPropertyInfo == JsonPropertyInfo.s_missingProperty)
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs]: I think the overlap is too small. If the implementation changes to do a full read on the second pass ([which is slower](https://github.com/dotnet/runtime/pull/33444#issuecomment-598546371) - I don't think we should go down this path), then we'd have very little code duplication here or in related methods

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs [hunk 12]
```diff
@@ -0,0 +1,694 @@
...
+                // Obtain the CLR value from the JSON and set the member.
+                if (state.Current.UseExtensionProperty)
+                {
+                    if (state.Current.JsonClassInfo.DataExtensionIsObject)
+                    {
+                        if (!jsonPropertyInfo.ReadJsonExtensionDataValue(ref state, ref reader, out object? extDataValue))
+                        {
+                            return false;
+                        }
+                        ((IDictionary<string, object>)state.Current.DataExtension!)[state.Current.JsonPropertyNameAsString!] = extDataValue!;
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs]: I don't think so - if the call is successful, then we'll fall out of the `if` block, end the property, and return `true`.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs [hunk 13]
```diff
@@ -0,0 +1,694 @@
...
+                    state.Current.FoundPropertiesAsync![state.Current.FoundPropertyCount++] = (
+                            jsonPropertyInfo,
+                            state.Current.ReaderState,
+                            reader.OriginalSpan.Slice(checked((int)reader.TokenStartIndex)).ToArray(),
+                            propertyNameArray);
+
+
+                    if (!reader.TrySkip())
+                    {
+                        return false;
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs]: I don't think so - when we reenter this method, the condition `state.Current.PropertyState < StackFramePropertyState.TryRead` will be true, and we'll end the property and exit:
 https://github.com/dotnet/runtime/blob/f23facb6348040c321cee4e0429c31d2faf2e959/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs#L576-L580

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs [hunk 14]
```diff
@@ -0,0 +1,694 @@
...
+                    }
+                }
+
+                if (state.Current.PropertyState < StackFramePropertyState.TryRead)
+                {
+                    state.Current.EndProperty();
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs]: Yes, we need to clear state we set when we first saw the property, including `PropertyState` so we can read the next property name.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs]: Is there a way to force tests down code paths where `reader.Read()` and `reader.TrySkip()` return `false`? I tried changing the `DefaultBufferSize` on `JsonSerializerOptions` with no luck.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs]: OK. So 0-25% depending.

A different approach, probably already considered if we can do a full second pass:
- Add a "CreateObject" virtual method to the base object converter or a new base class (like we do for collections, but need to pass the reader)
- The CreateObject for the ctor-supporting converter would do the first pass and create the object
- The base class then does the second pass (no code duplication). Likely we'd have to ensure property setters aren't called if they were passed into the ctor -- we can set ShouldDeserialize to false on those JsonPropertyInfos when the JsonClassInfo is created.

The benefit is avoiding duplicated lots of code.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs]: Good question. We should discuss options off-line here. By changing DefaultBufferSize to 1 (really 16) and adding whitespace between JSON tokens we should be able to force these.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs (hunk 3 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: Are both values are set for the given ClassInfo instance before it is added to the ClassInfo cache? If so, then we should be OK w.r.t. threading.

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: Yes, they are.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs]: Per offline discussion this is only possible if we "read-ahead" or "stitch" the previously read data from the first pass. This is likely not feasible.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ConstructorArgumentState.cs [hunk 15]
```diff
@@ -0,0 +1,53 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    /// // Holds relevant state when deserializing objects with parameterized constructors.
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ConstructorArgumentState.cs]: nit: extra //

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ConstructorArgumentState.cs [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+        public List<ParameterRef>? ParameterRefCache;
+
+        // The starting position of an object property when supporting continuation. This is cached for
+        // each property and used to deserialize the property after the object is constructed.
+        public JsonReaderState ReaderState;
+
+        public void Reset()
+        {
+            ParameterIndex = 0;
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ConstructorArgumentState.cs]: Can we do a `this = default` here? Looks like `ReaderState` and `JsonParameterInfo` weren't cleared -- should they have been?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ConstructorArgumentState.cs [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+using System.Collections.Generic;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    /// // Holds relevant state when deserializing objects with parameterized constructors.
+    /// Lives on the current ReadStackFrame.
+    /// </summary>
+    internal struct ConstructorArgumentState
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ConstructorArgumentState.cs]: This is much cleaner. Thanks

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs]: As part of mitigating perf regression seen for type graphs with numerous objects to deserialize, I've done some refactoring which reduces a ton of the duplicated code.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ConstructorArgumentState.cs (hunk 16 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ConstructorArgumentState.cs]: They should have been, thanks. This method is removed.

File: src/libraries/System.Text.Json/tests/Serialization/DeserializationWrapper.cs [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+                    {
+                        using (MemoryStream stream = new MemoryStream(Encoding.UTF8.GetBytes(json)))
+                        {
+                            return await JsonSerializer.DeserializeAsync<T>(stream, options);
+                        }
+                    }).Result;
+                }
+                catch (Exception e)
+                {
+                    throw e.InnerException;
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/DeserializationWrapper.cs]: Is this because of Task throwing `AggregateException`? Is so should we be catching that instead?

@stephentoub what is the best practice here? Should we be using ValueTask instead? (which won't throw the AggregateException AFAIK)

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests.Stream.cs [hunk 19]
```diff
@@ -2,111 +2,135 @@
...
 
-                var obj = await JsonSerializer.DeserializeAsync(stream, type, options);
-                ((ITestClass)obj).Verify();
+                    var obj = await JsonSerializer.DeserializeAsync<T>(stream, options);
+                    ((ITestClass)obj).Verify();
+                }
             }
+
+            // Simple models can be deserialized.
+            Task.Run(async () => await RunTest<Parameterized_IndexViewModel_Immutable>(Parameterized_IndexViewModel_Immutable.s_data));
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests.Stream.cs]: Using `Task.Run` like this will "fire and forget" since the task is not awaited anywhere else. This could cause random exceptions later. The inner `await` doesn't change the semantics that AFAIK.

If you want to run all of these async, get the task from each and call `Task.WhenAll` on the set. That will ensure the tasks finish before this test returns.

File: src/libraries/System.Text.Json/tests/Serialization/DeserializationWrapper.cs (hunk 18 above)
Comment (stephentoub) [reply src/libraries/System.Text.Json/tests/Serialization/DeserializationWrapper.cs]:
```C#
return Task.Run(async () =>
{
    ...
}).GetAwaiter().GetResult();
```

**Review by steveharter — APPROVED** [review]

**Comment by layomia** [comment]
Test failure unrelated - [System.Xml.RW.XmlConvert.Tests](https://dev.azure.com/dnceng/public/_build/results?buildId=567199&view=logs&j=76fc711c-4822-538a-1117-db98382fba3d&t=aff8ddcf-c443-5f84-814a-d4c702f1006b.).
